vscode/src/player/Subtitles.ts

```diff
@@ -0,0 +1,151 @@
+ */
+const TIME_UNIT: number = 1000;
+
+/**
+ * Adjustment multiplier for timers to execute sooner than later.
+ */
+const ADJ_MUL: number = 0.98;
+
+/**
+ * Class handles loading and manipulation of the subtitles.
+ */
+export default class Subtitles {
+  private oc: vscode.OutputChannel;
+  private cuesDict: Object = {};
+  private startTimeUnit: number;
+  private timerRef: NodeJS.Timer;
+
+  /**
+   * Create Output Channel.
+   */
+  constructor() {
+    this.oc = vscode.window.createOutputChannel('Codio Subtitles');
+  }
+
+  /**
+   * Load given subtitles file and create dictionary.
+   * @param filePath Subtitles file to load.
+   */
+  async load(filePath: string): Promise<void> {
+    const uri = vscode.Uri.file(filePath);
+
+    try {
+      const srtData = (await vscode.workspace.openTextDocument(uri)).getText();
+      const cues = parser.fromSrt(srtData, 'ms');
+      if (!cues.length) {
+        throw new Error('No subtitles found.');
+      }
+
+      this.createCuesDict(cues);
+      this.display('Codio Subtitles Loaded');
+    } catch (error) {
+      console.error('Subtitles Load', error.message);
+      this.play = () => { };
+    }
+  }
+
+  /**
+   * Create a dictionary of cue start times and format strings.
+   * @param cues Array of cue objects to parse.
+   */
+  private createCuesDict(cues: Cue[]): void {
+    let tu: number;
+    for (const cue of cues) {
+      tu = this.getRestriction(cue.startTime);
+      this.cuesDict[tu] = this.getCueStr(cue);
+    }
+  }
+
+  /**
+   * Display given string on the output channel.
+   * @param str String to display.
+   */
+  private display(str: string) {
+    this.oc.show(true);
+    this.oc.appendLine(str);
+  }
+
+  /**
+   * Return given milliseconds to a restricted time unit.
+   * @param ms Milliseconds to restrict.
+   * @param tu Optional: Time unit to restrict to.
+   */
+  private getRestriction(ms: number, tu = TIME_UNIT): number {
+    return Math.floor(ms - (ms % tu));
+  }
+
+  /**
+   * Format given cue into a string.
+   * @param cue Cue to format into a string.
+   */
+  private getCueStr(cue: Cue): string {
+    return parser.msToSrt(cue.startTime) + '\t' + cue.text;
+  }
+
+  /**
+   * Play subtitles from given milliseconds.
+   * @param timeMs Milliseconds into subtitles to play.
+   */
+  play(timeMs = 0): void {
+    this.stop();
+
+    if (!timeMs) {
+      this.oc.clear();
+      this.display('Codio Subtitles Starting...');
+    }
+
+    this.startTimeUnit = this.getRestriction(timeMs);
+    this.handleTimeUnit(this.startTimeUnit);
+
+    this.start();
+  }
+
+  /**
+   * Display cue string at given time unit if available.
+   * @param tu Time unit to use to find cue string.
+   */
+  private handleTimeUnit(tu: number): void {
+    if (this.cuesDict[tu]) {
+      this.display(this.cuesDict[tu]);
+    }
+  }
+
+  /**
+   * Start to display at certain time units.
+   */
+  private start(): void {
+    this.timerRef = setInterval(() => {
```

Comment:
This is a bit foreign to me, but I guess it has to do with how the parser works. 
In our ordinary timeline, we simply setTimeout to the next `time` which could be the next '`start time` of every unit.  Why do it differently here? Is this because it is how the data structure supported by the parser is expected to be used?

Reply:
Precisely. Specifically in [createCuesDict](https://github.com/wix-incubator/codio/pull/39/files#diff-0fef0568e8c0de86bf3edfd6a1071192b97b7ec270fcda477c3dfecf2dc34bfcR66). Because `setTimeout` is not exact the audio and the subtitles would eventually be out of sync. TIME_UNIT and ADJ_MUL were adjusted to combat this and the resulting values were satisfactory in my tests.

Reply:
setTimeout is not exact, but we solve the "dragging out-of-sync problem" in our runThroughTimeline function by calculating the diff between Date.now() and the designated time the event should be fired on. We do this before every event, so even if the previous event was delayed, the next setTimeout is not affected by it. This seems to work well enough for our core event logic, so we can use the same function in subtitles. See the `runThroughTimeline` function in the `timeline` file

Reply:
I will take a look and see if it satisfies the issue.

Reply:
Thank you for pointing out this issue. I believe I was trying different ways to solve this and ended up with a non-standard approach. It is now mathematically correct in syncing up and less code too!